### PR TITLE
build: allow disabling snapshots for dev

### LINKED ISSUE #19399: Ability to create and load a snapshot without rebuilding the whole binary
In an effort to improve iteration time when working **only** on JavaScript code,
we should consider having a way to rebuild and load the startup snapshot without
having to rebuild the whole binary.

This would be done by a combination of a Cargo feature flag (or env variable) to
instruct the `cargo build` process to not include the snapshot in the binary, and
a separate script that would be able to build the snapshot on demand.

This might require building a separate binary that would have a sole purpose of
building the snapshot (akin to V8's `mksnapshot` binary used to produce initial
snapshot with JS built-ins).

There is a challenge related to the fact that we currently snapshot "ops" that are
included in the binary, but it doesn't seem insurmountable.

### PR DESCRIPTION
**Pull request by nayeemrmn**
Closes #19399 (running without snapshots at all was suggested as an alternative solution).

Adds a `__runtime_js_sources` pseudo-private feature to load extension JS sources at runtime for faster development, instead of building and loading snapshots or embedding sources in the binary. Will only work in a development environment obviously.

Try running `cargo test --features __runtime_js_sources integration::node_unit_tests::os_test`. Then break some behaviour in `ext/node/polyfills/os.ts` e.g. make `function cpus() {}` return an empty array, and run it again. Fix and then run again. No more build time in between.

Blocked by #20043.

cc @bartlomieju

### CODE REVIEW

**Comment by bartlomieju** [comment]
@nayeemrmn fantastic! I think I managed to find an edge case here.

If I do:
```
$ cargo build --features __runtime_js_sources
```
And then:
```
$ ./target/debug/deno
```
I get:
```
./target/debug/deno

============================================================
Deno has panicked. This is a bug in Deno. Please report this
at https://github.com/denoland/deno/issues/new.
If you can reliably reproduce this panic, include the
reproduction steps and re-run with the RUST_BACKTRACE=1 env
var set and include the backtrace in your report.

Platform: macos aarch64
Version: 1.36.0
Args: ["./target/debug/deno"]

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: NotPresent', ext/node/lib.rs:136:6
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

The panic doesn't appear if I do `cargo run --features __runtime_js_sources`. Any ideas?

**Comment by nayeemrmn** [comment]
Strange, I'm not getting that error. From a quick glance at https://github.com/denoland/deno/blob/5311f69bbbd169726330ddae20ce119624b8e8ca/ext/node/lib.rs#L135-L136 why is it reading `TARGET` at runtime instead of `env!()`?

**Comment by bartlomieju** [comment]
> Strange, I'm not getting that error. From a quick glance at
> 
> https://github.com/denoland/deno/blob/5311f69bbbd169726330ddae20ce119624b8e8ca/ext/node/lib.rs#L135-L136
> 
> why is it reading `TARGET` at runtime instead of `env!()`?

This is the diagnostic I get if I try to use `env!`:
```
environment variable `TARGET` not defined at compile time
Cargo sets build script variables at run time. Use `std::env::var("TARGET")` instead
```

**Comment by bartlomieju** [comment]
It's strange you don't get that error - this op is called in 3 separate places by `ext/node` so you should definitely be calling it if executing some `node:` built-in module or `npm:` package.

**Comment by bartlomieju** [comment]
> Okay it's because the TARGET is defined in `runtime/build.rs`,`deno_node` is build before that.

We should probably move all of that to `deno_core` and have build info attached there to `Deno.core.build`. We are currently using these env vars in several places :/

**Comment by bartlomieju** [comment]
@nayeemrmn this is absolutely fantastic. I've been playing with it for half an hour now and it's a game changer in terms of working on the JS APIs. I love how few changes you had to do, I'll release `deno_core` tonight after Matt finishes some stuff so we can land it ASAP.

**Comment by marvinhagemeister** [comment]
Just wanted to say that this is fantastic feature! Really looking forward to seeing this sometime in Deno 👍

**Comment by nayeemrmn** [comment]
> $ ./target/debug/deno

By the way I could reproduce the error with `target/debug/deno`, using `cargo run` seems to initialise the `cargo:rustc-env` values on startup (for me only, still strange). Anyway I added `target-lexicon` from bytecodealliance to get the target triple at compile time.

> Just wanted to say that this is fantastic feature! Really looking forward to seeing this sometime in Deno 👍

Thanks!

File: ext/node/lib.rs [hunk 1]
```diff
@@ -132,12 +132,7 @@ pub static NODE_ENV_VAR_ALLOWLIST: Lazy<HashSet<String>> = Lazy::new(|| {
 
 #[op]
 fn op_node_build_os() -> String {
-  std::env::var("TARGET")
-    .unwrap()
-    .split('-')
-    .nth(2)
-    .unwrap()
-    .to_string()
+  target_lexicon::HOST.operating_system.to_string()
```
Comment (nayeemrmn) [inline ext/node/lib.rs:135]: Actually this is the host not the target 🤔

**Comment by sigmaSd** [comment]
does this have any effect on cross compiling, as in can I now crosscompile by disabling snapshotting ?

**Comment by nayeemrmn** [comment]
> does this have any effect on cross compiling, as in can I now crosscompile by disabling snapshotting ?

No, this snapshot-less mode creates a binary that only works on the development machine it was built on. That would require an alternative snapshot-less mode which bundles all the sources in the binary.

**Review by bartlomieju — APPROVED** [review]
LGTM, fantastic work @nayeemrmn, this will make working on JS/TS polyfills way faster! Kudos 🚀

**Pull request merged**